zerver/lib/exceptions.py

```diff
@@ -177,7 +176,7 @@ def __init__(self, include_statement: str) -> None:
     def msg_format() -> str:
         return _("Invalid markdown include statement: {include_statement}")
 
-class RateLimited(PermissionDenied):
+class RateLimited(Exception):
     def __init__(self, msg: str="") -> None:
         super().__init__(msg)
 
```

Comment:
If we're doing this, we can remove the Union mypy hack in `zerver/middleware.py`.

Also, we should make sure that this returns a reasonable HTTP status code; I suspect that it isn't doing so correctly.

Reply:
If I understand what you mean, I believe test_external.py tests for this in:
```
    def test_hit_ratelimits(self) -> None:
        user = self.example_user('cordelia')
        email = user.email
        clear_history(RateLimitedUser(user))

        start_time = time.time()
        for i in range(6):
            with mock.patch('time.time', return_value=(start_time + i * 0.1)):
                result = self.send_api_message(email, "some stuff %s" % (i,))

        self.assertEqual(result.status_code, 429)
        json = result.json()
        self.assertEqual(json.get("result"), "error")
        self.assertIn("API usage exceeded rate limit", json.get("msg"))
        self.assertEqual(json.get('retry-after'), 0.5)
        self.assertTrue('Retry-After' in result)
        self.assertEqual(result['Retry-After'], '0.5')
```

RateLimitMiddleware captures the RateLimited exception and raises the response with the 429 status code:
```
            resp = json_error(
                _("API usage exceeded rate limit"),
                data={'retry-after': request._ratelimit[entity_type].secs_to_freedom},
                status=429
            )
```

Reply:
This comment still applies

Reply:
OK, sounds good